### ISSUE TITLE
cat: os error 3 on MSYS2 package

### ISSUE THREAD

**Issue by ognevnydemon** [uutils/coreutils#5417]
after installing [MSYS2 provided package](https://packages.msys2.org/base/mingw-w64-uutils-coreutils) (cat's path is `C:\msys64\clang64\bin\cat.exe`) I recieve `os error 3` after every start for path `'/proc/registry/HKEY_CURRENT_USER/Software/Microsoft/Windows NT/CurrentVersion/Windows/Device'`

**Comment by tertsdiepraam** [comment]
So os error 3 is path not found right? Could that have something to do with using a unix-style path? Could you give us the full error that you see?

**Comment by tertsdiepraam** [comment]
Is `/proc` valid on Windows? I don't think I fully understand your setup here. You seem to be using the Windows version of `cat` on a Unix path, is that correct?

**Comment by ognevnydemon** [comment]
> Member



> You seem to be using the Windows version of `cat` on a Unix path, is that correct?

it's built with [this script](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-uutils-coreutils/PKGBUILD): just `make PROFILE=release` for build. mostly MSYS2 packages compiled for mingw targets support both unix- and windows-style path.

> Is /proc valid on Windows? Windows does not really have a concept of a root directory right?

can you help, @MehdiChinoune? I think you can provide a bit more information about it

**Comment by tertsdiepraam** [comment]
I see, so you want uutils on Windows to support both types of paths? Based on https://www.msys2.org/wiki/MSYS2-introduction/, I think that the path you are trying to access is a virtual filesystem from MSYS2. If I understand that correctly, it would be something very specific to only MSYS2.

So, I don't think uutils should support it by default. One possibility is for MSYS2 to patch uutils and normalize all paths to Windows style before passing to uutils, but that's not easy to do for all utils. If you (or someone from MSYS2) can tell us how they usually do it, we can of course discuss that further.

**Comment by ognevnydemon** [comment]
> I see, so you want uutils on Windows to support both types of paths? Based on https://www.msys2.org/wiki/MSYS2-introduction/, I think that the path you are trying to access is a virtual filesystem from MSYS2. If I understand that correctly, it would be something very specific to only MSYS2.
> 
> So, I don't think uutils should support it by default. One possibility is for MSYS2 to patch uutils and normalize all paths to Windows style before passing to uutils, but that's not easy to do for all utils. If you (or someone from MSYS2) can tell us how they usually do it, we can of course discuss that further.
> 

there is a base util `cygpath` that convert path to needed style. but I personally have no experience with it

**Comment by oech3** [comment]
I know `C:\` is `/c/`. But I think uutils/coreutils cannot support device file even patch was accepted at here.

**Comment by ognevny** [comment]
> `'/proc/registry/HKEY_CURRENT_USER/Software/Microsoft/Windows NT/CurrentVersion/Windows/Device'`

cygwin path conversion should turn this into `${msys2_installation_dir}/proc/registry/HKEY_CURRENT_USER/Software/Microsoft/Windows NT/CurrentVersion/Windows/Device`. e.g. MSYS2 is under `C:\msys64` so the full path is `C:\msys64\proc\registry\HKEY_CURRENT_USER\Software\Microsoft\Windows NT\CurrentVersion\Windows\Device`. in gnu coreutils for cygwin it's handled inside of cygwin runtime IIRC

**Comment by ChrisDenton** [comment]
I think this will be helped by the [cygwin target](https://doc.rust-lang.org/nightly/rustc/platform-support/x86_64-pc-cygwin.html). It's currently very new and not fully implemented yet so if anyone reading this wants to help out then I'm sure the maintainers would appreciate it.

**Comment by ognevny** [comment]
> I think this will be helped by the [cygwin target](https://doc.rust-lang.org/nightly/rustc/platform-support/x86_64-pc-cygwin.html).

absolutely! that's my point of discussion we had downstream... better have cygwin package with no prefix in binaries name then trying to adapt MINGW alternatives which are not used for msys2's internal scripts

**Comment by oech3** [comment]
>@tertsdiepraam 
So, I don't think uutils should support it by default.

How about adding `#[cfg(target_os = ...)]` for cygwin at this repository (not sure if it is enough)?

**Comment by oech3** [comment]
https://github.com/nix-rust/nix/commit/975111f2a79b69c26421e0979c50eb62cee336e7

**Comment by ognevny** [comment]
well, but the package still fails to build under cygwin due to limited features of libc crate

**Comment by ognevny** [comment]
well, but the package still fails to build under cygwin due to limited features of libc crate

**Comment by ognevny** [comment]
on Cygwin host the issue doesn't appear!

**Comment by oech3** [comment]
I'm still interested in native binaries. So I want to open a new issue for
`feat_cygpath` on MinGW/MSVC. OK?

**Comment by ognevny** [comment]
> I'm still interested in native binaries. So I want to open a new issue for
> `feat_cygpath` on MinGW/MSVC. OK?

if you wish to implement it... why not?
